Fix the RedirectHandler (in Mvc.Testing) for relative URIs

0xced opened the pull request:
# Fix the RedirectHandler (in Mvc.Testing) for relative URIs

- [x] You've read the [Contributor Guide](https://github.com/dotnet/aspnetcore/blob/main/CONTRIBUTING.md) and [Code of Conduct](https://github.com/dotnet/aspnetcore/blob/main/CODE-OF-CONDUCT.md).
- [x] You've included unit or integration tests for your change, where applicable.
- [x] You've included inline docs for your change, where applicable.
- [ ] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.

## Description

This was a case of double bugs cancelling each other. The `Testing/RedirectHandler/Redirect303` route was redirecting to `Testing/Builder` instead of `/Testing/Builder`, hiding the fact that the relative URL in the RedirectHandler was not handled properly.

A new test covering relative URLs has been added.

0xced committed:
Fix the RedirectHandler (in Mvc.Testing) for relative URIs

This was a case of double bugs cancelling each other. The `Testing/RedirectHandler/Redirect303` route was redirecting to `Testing/Builder` instead of `/Testing/Builder`, hiding the fact that the relative URL in the RedirectHandler was not handled properly.

A new test covering relative URLs has been added.

javiercn commented:
@captainsafia @brunolins16 can you take a look?

javiercn commented:
@0xced for future reference, please file an issue no matter how small the change before sending a PR explaining the issue in detail. It helps us get context on what's going on, what the current behavior is and what why the new behavior is better/expected/desired/correct.

captainsafia commented:
> It helps us get context on what's going on, what the current behavior is and what why the new behavior is better/expected/desired/correct.

Can you share more details on this in the PR or a new issue? I'm having trouble reconciling the issue description above with the change to the test app [here](https://github.com/dotnet/aspnetcore/pull/41742/files#diff-29bfbb1453c63f6f30d6ef998c10bb94f79a4895f6e171d067d6582d568f2a81R78).

0xced commented:
The `/Testing/RedirectHandler/Redirect303` route was redirecting to `Testing/Builder`. If you use `curl -L` or your browser, you are redirected to `/Testing/RedirectHandler/Testing/Builder` because the redirect location is relative. The test `TestingInfrastructure_RedirectHandlerFollowsStatusCode303` should fail because `/Testing/RedirectHandler/Testing/Builder` is a 404 and not a 200. But the test did succeed because of how the redirect location was previously computed, stripping the `/Testing/RedirectHandler` part:
```csharp
location = new Uri(new Uri(requestUri.GetLeftPart(UriPartial.Authority)), location);
```

So I fixed the `RedirectHandlerStatusCode303()` route implementation to redirect to `/Testing/Builder` with an absolute path. That's what `RedirectToAction` which is used in all other tests does by the way.

Finally, I fixed the construction of the redirect location and added a new test (`TestingInfrastructure_RedirectHandlerHandlesRelativeLocation`) to explicitly cover a redirection to a relative path: `/Testing/RedirectHandler/Relative/` redirects to `Ok` which must end up being `/Testing/RedirectHandler/Relative/Ok`.
```csharp
location = new Uri(requestUri, location);
```

Now I'm left wondering: what is the purpose of the `RedirectHandler` class in the testing infrastructure? I mean, the standard `HttpClientHandler` has an `AllowAutoRedirect` property and properly handles redirection, right?

brunolins16 commented on src/Mvc/Mvc.Testing/src/Handlers/RedirectHandler.cs:
```diff
@@ -144,9 +144,7 @@ private static async Task<(Stream originalBody, Stream copy)> CopyBody(HttpReque
-                location = new Uri(
-                    new Uri(requestUri.GetLeftPart(UriPartial.Authority)),
-                    location);
+                location = new Uri(requestUri, location);
```
That change looks correct for me and will be consistent with what we have in the "real" `HttpClient`

https://github.com/dotnet/runtime/blob/56e821243d92f20a6abf7ca61288221f4d164dd1/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/RedirectHandler.cs#L112

My only concern if that will impact tests and we should add a **Breaking change announcement**.

brunolins16 commented:
> Now I'm left wondering: what is the purpose of the `RedirectHandler` class in the testing infrastructure? I mean, the standard `HttpClientHandler` has an `AllowAutoRedirect` property and properly handles redirection, right?

I did not investigate but we are chaining the handlers and that might not be possible with the real `RedirectHandler`.

brunolins16 commented:
> Now I'm left wondering: what is the purpose of the `RedirectHandler` class in the testing infrastructure? I mean, the standard `HttpClientHandler` has an `AllowAutoRedirect` property and properly handles redirection, right?

I did not investigate but we are chaining the handlers and that might not be possible with the real `RedirectHandler`.

brunolins16 reviewed (approved):
@captainsafia do you have any additional concerns about this change?

captainsafia reviewed (approved):
> My only concern if that will impact tests and we should add a Breaking change announcement.

I don't think this is a breaking change. It seems like a case of poorly written tests hiding the fact that there was a bug in our code and it seems strange that someone would rely on this buggy behavior.

brunolins16 commented:
@0xced Thanks for your contribution!
